From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Tue, 13 Oct 2026 00:57:04 -0700
Subject: [PATCH] jscript: call and apply on a host function ask for a method
 call only

In the legacy (non-HTML) mode, Function.prototype.call and apply pass
DISPATCH_METHOD without DISPATCH_PROPERTYGET to a host function, as the
conformance test records for native. The HTML/ES5 modes keep both flags.

diff --git a/dlls/jscript/dispex.c b/dlls/jscript/dispex.c
index d5ed3ad..52973f7 100644
--- a/dlls/jscript/dispex.c
+++ b/dlls/jscript/dispex.c
@@ -2773,6 +2773,7 @@ HRESULT disp_call_value_with_caller(script_ctx_t *ctx, IDispatch *disp, jsval_t
     DISPPARAMS dp;
     unsigned i;
     HRESULT hres = S_OK;
+    WORD noget;
 
     static DISPID this_id = DISPID_THIS;
 
@@ -2780,13 +2781,15 @@ HRESULT disp_call_value_with_caller(script_ctx_t *ctx, IDispatch *disp, jsval_t
 
     jsdisp = iface_to_jsdisp(disp);
     if(jsdisp && jsdisp->ctx == ctx) {
-        hres = jsdisp_call_value(jsdisp, vthis, flags, argc, argv, r);
+        hres = jsdisp_call_value(jsdisp, vthis, flags & ~DISPATCH_JSCRIPT_NOPROPGET, argc, argv, r);
         jsdisp_release(jsdisp);
         return hres;
     }
     if(jsdisp)
         jsdisp_release(jsdisp);
 
+    noget = flags & DISPATCH_JSCRIPT_NOPROPGET;
+
     if(is_object_instance(vthis) && (ctx->version < SCRIPTLANGUAGEVERSION_ES5 ||
        ((jsdisp = to_jsdisp(get_object(vthis))) && is_class(jsdisp, JSCLASS_OBJECT))))
         jsthis = get_object(vthis);
@@ -2794,7 +2797,11 @@ HRESULT disp_call_value_with_caller(script_ctx_t *ctx, IDispatch *disp, jsval_t
         jsthis = NULL;
 
     flags &= ~DISPATCH_JSCRIPT_INTERNAL_MASK;
+    #ifdef SG_MUTANT_JS_NOGET
     if(r && argc && flags == DISPATCH_METHOD)
+#else
+    if(r && argc && flags == DISPATCH_METHOD && !noget)
+#endif
         flags |= DISPATCH_PROPERTYGET;
 
     if(jsthis) {
diff --git a/dlls/jscript/function.c b/dlls/jscript/function.c
index ea8a1da..e080fe7 100644
--- a/dlls/jscript/function.c
+++ b/dlls/jscript/function.c
@@ -509,7 +509,7 @@ static HRESULT Function_apply(script_ctx_t *ctx, jsval_t vthis, WORD flags, unsi
             hres = function->vtbl->call(ctx, function, this_val, flags, cnt, args, r);
         }else {
             jsval_t res;
-            hres = disp_call_value(ctx, get_object(vthis), this_val, DISPATCH_METHOD, cnt, args, &res);
+            hres = disp_call_value(ctx, get_object(vthis), this_val, DISPATCH_METHOD | (ctx->html_mode ? 0 : DISPATCH_JSCRIPT_NOPROPGET), cnt, args, &res);
             if(SUCCEEDED(hres)) {
                 if(r)
                     *r = res;
diff --git a/dlls/jscript/jscript.h b/dlls/jscript/jscript.h
index dab8d42..6bfbf35 100644
--- a/dlls/jscript/jscript.h
+++ b/dlls/jscript/jscript.h
@@ -96,7 +96,9 @@ HRESULT get_dispatch_typeinfo(ITypeInfo**);
  * of execution.
  */
 #define DISPATCH_JSCRIPT_CALLEREXECSSOURCE  0x8000
-#define DISPATCH_JSCRIPT_INTERNAL_MASK      DISPATCH_JSCRIPT_CALLEREXECSSOURCE
+/* The call of a host function through Function.prototype.call/apply does not ask for a property value. */
+#define DISPATCH_JSCRIPT_NOPROPGET          0x2000
+#define DISPATCH_JSCRIPT_INTERNAL_MASK      (DISPATCH_JSCRIPT_CALLEREXECSSOURCE | DISPATCH_JSCRIPT_NOPROPGET)
 
 /* NOTE: Keep in sync with names in Object.toString implementation */
 typedef enum {
diff --git a/dlls/jscript/tests/run.c b/dlls/jscript/tests/run.c
index ee8246f..e7c6c89 100644
--- a/dlls/jscript/tests/run.c
+++ b/dlls/jscript/tests/run.c
@@ -782,7 +782,7 @@ static HRESULT WINAPI dispexFunc_InvokeEx(IDispatchEx *iface, DISPID id, LCID lc
 
         if(V_BOOL(pdp->rgvarg+1))
             /* NOTE: If called by Function.apply(), native doesn't set DISPATCH_PROPERTYGET flag. */
-            todo_wine ok(wFlags == DISPATCH_METHOD, "wFlags = %x\n", wFlags);
+            ok(wFlags == DISPATCH_METHOD, "wFlags = %x\n", wFlags);
         else
             ok(wFlags == (DISPATCH_PROPERTYGET|DISPATCH_METHOD), "wFlags = %x\n", wFlags);
 
@@ -870,7 +870,7 @@ static HRESULT WINAPI pureDisp_Invoke(IDispatchEx *iface, DISPID dispIdMember, R
         ok(V_VT(pdp->rgvarg) == VT_BOOL, "V_VT(pdp->rgvarg) = %d\n", V_VT(pdp->rgvarg));
 
         if(V_BOOL(pdp->rgvarg))
-            todo_wine ok(wFlags == DISPATCH_METHOD, "wFlags = %x\n", wFlags);
+            ok(wFlags == DISPATCH_METHOD, "wFlags = %x\n", wFlags);
         else
             ok(wFlags == (DISPATCH_PROPERTYGET|DISPATCH_METHOD), "wFlags = %x\n", wFlags);
 
